doc/add_datafit.rst

```diff
@@ -0,0 +1,120 @@
+:orphan:
+
+How to add a custom datafit
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _how:
+
+Motivated by generalized linear models but not limited to it, ``skglm`` solves problems of the form
+
+.. math::
+      \hat{\beta} \in
+      \arg\min_{\beta \in \mathbb{R}^p}
+      F(X\beta) + \Omega(\beta)
+      := \sum_{i=1}^n f_i(X_i^{\top}\beta) + \sum_{j=1}^p \Omega_j(\beta_j)
+      \enspace .
+
+
+Here, :math:`X \in \mathbb{R}^{n \times p}` denotes the design matrix with :math:`n` samples and :math:`p` features,
+and :math:`\beta \in \mathbb{R}^p` is the coefficient vector.
+
+``skglm`` can solve any problems of this form with arbitrary smooth datafit :math:`F` and arbitrary proximable penalty :math:`\Omega`,
+by defining two classes: a ``Penalty`` and a ``Datafit``.
+
+They can then be passed to a :class:`~skglm.GeneralizedLinearEstimator`.
+
+.. code-block:: python
+
+   clf = GeneralizedLinearEstimator(
+      MyDatafit(),
+      MyPenalty(),
+   )
+
+
+A ``Datafit`` is a jitclass which must inherit from the ``BaseDatafit`` class:
+
+.. literalinclude:: ../skglm/datafits/base.py
+   :pyobject: BaseDatafit
+
+
+To define a custom datafit, you need to implement the methods declared in the ``BaseDatafit`` class (value, gradient, gradient when the design matrix is sparse).
```

Comment:
(value, gradient, gradient when the design matrix is sparse) ---> members that ends with ``_sparse`` are optional. They are required only to add support for sparse datasets.